Add RH Cloud - Inventory support

mirekdlugosz opened the pull request:
Adding entity and view for RH Cloud - Inventory plugin, available in Satellite under "RH Cloud" -> "Inventory Upload" menu item.
In Satellite before 6.8 snap 5 you need to install plugin manually: `satellite-installer --enable-foreman-plugin-inventory-upload` or `satellite-installer --enable-foreman-plugin-rh-cloud` (depending on Satellite version).

Currently, there are no Robottelo tests that use these. I will be adding them later, but I'm going to need some changes in Robottelo first.

Also, I will be backporting these to all branches for supported Satellite versions, once I complete entire workflow.

tests:
```bash
$ cat tests/foreman/ui/test_cloudinventory.py 
from airgun.session import Session


def test_cloud():
    with Session(user='admin', password='changeme') as session:
        print(session.cloudinventory.read())
        print(session.cloudinventory.read('Default Organization'))
        session.cloudinventory.update({
            'allow_auto_upload': False,
            'obfuscate_host_names': False,
        })
        session.cloudinventory.generate_report('Default Organization')
        report_path = session.cloudinventory.download_report('Default Organization')
        print(report_path)
        print(session.cloudinventory.read('Default Organization'))
```
```bash
$ pytest -s -k cloud tests/foreman/ui/test_cloudinventory.py 
===================================================================================== test session starts =====================================================================================
platform linux -- Python 3.8.3, pytest-4.6.3, py-1.8.1, pluggy-0.13.1
shared_function enabled - OFF - scope:  - storage: file
rootdir: /home/mzalewsk/sources/robottelo
plugins: cov-2.10.0, mock-1.10.4, forked-1.2.0, services-1.3.1, xdist-1.32.0
collecting ... 2020-07-09 19:52:22 - conftest - DEBUG - Collected 1 test cases
satellite-6.8.0-0.6.beta.el7sat.noarch

2020-07-09 21:52:26 - robottelo.ssh - DEBUG - Destroyed Paramiko client 0x7f0375300c70
2020-07-09 21:52:26 - robottelo.host_info - DEBUG - Host Satellite version: 6.8
2020-07-09 21:52:26 - robottelo.helpers - DEBUG - Generated file bz_cache.json with BZ collect data
collected 1 item
tests/foreman/ui/test_cloudinventory.py 2020-07-09 21:52:26 - robottelo - DEBUG - Started Test: test_cloudinventory.py/test_cloud

{'menu': ['', 'Configure', 'Default Organization', 'Default Location'],
 'flash': [],
 'current_user': 'Admin User',
 'account_menu': 'Admin User',
 'logout': 'Log Out',
 'title': 'Red Hat Inventory Uploads',
 'searchbox': '',
 'allow_auto_upload': True,
 'obfuscate_host_names': True,
 'inventory_list': {'Default Organization': {'generating': {'process': 'Exit Code: pid 35823 exit 0',
                                                            'restart': 'Restart',
                                                            'terminal': 'Must specify either portal_user or organization_id I, [2020-07-09T15:41:38.292573 #35826] INFO -- : Started generating hosts report in /tmp/d20200709-35826-jsi3l8 I, [2020-07-09T15:41:38.339356 #35826] INFO -- : Report generation finished I, [2020-07-09T15:41:38.339445 #35826] INFO -- : Archiving generated report I, [2020-07-09T15:41:38.344598 #35826] INFO -- : tar: ./ ./metadata.json I, [2020-07-09T15:41:38.344810 #35826] INFO -- : Report archived successfully Successfully generated /var/lib/foreman/red_hat_inventory/generated_reports/report_for_1.tar.gz for organization id 1',
                                                            'scheduled_run': 'Next run: in 13 hours.'},
                                             'uploading': {'process': 'Exit Code: pid 35868 exit 0',
                                                           'download_report': 'Download Report',
                                                           'terminal': '* About to connect() to qa.cloud.redhat.com port 443 (#0) * Trying 10.0.19.25... * Connected to qa.cloud.redhat.com (10.0.19.25) port 443 (#0) *  Done: /var/lib/foreman/red_hat_inventory/uploads/report_for_1.tar.gz Uploaded files moved to done/ folder'},
                                             'status': 'idle',
                                             'allow_auto_upload': True,
                                             'obfuscate_host_names': True},
                    'łatever organization': {'generating': {'process': 'Exit Code: pid 48292 exit 0',
                                                            'restart': 'Restart',
                                                            'terminal': 'skipped',
                                                            'scheduled_run': 'Next run: in 13 hours.'},
                                             'uploading': {'process': 'Exit Code:', 'download_report': 'Download Report', 'terminal': ''},
                                             'status': 'idle',
                                             'allow_auto_upload': True,
                                             'obfuscate_host_names': True},
                    '~!@#$%^&*()_+-=;:\'"[]{},./<>?': {'generating': {'process': 'Exit Code: pid 48283 exit 0',
                                                                      'restart': 'Restart',
                                                                      'terminal': 'skipped',
                                                                      'scheduled_run': 'Next run: in 13 hours.'},
                                                       'uploading': {'process': 'Exit Code:', 'download_report': 'Download Report', 'terminal': ''},
                                                       'status': 'idle',
                                                       'allow_auto_upload': True,
                                                       'obfuscate_host_names': True},
                    'QnAbES': {'generating': {'process': 'Exit Code:', 'restart': 'Restart', 'terminal': '', 'scheduled_run': 'Next run: in 13 hours.'},
                               'uploading': {'process': 'Exit Code:', 'download_report': 'Download Report', 'terminal': ''},
                               'status': 'idle',
                               'allow_auto_upload': True,
                               'obfuscate_host_names': True},
                    'tzlxBfON': {'generating': {'process': 'Exit Code: pid 12089 exit 0',
                                                'restart': 'Restart',
                                                'terminal': 'skipped',
                                                'scheduled_run': 'Next run: in 13 hours.'},
                                 'uploading': {'process': 'Exit Code:', 'download_report': 'Download Report', 'terminal': ''},
                                 'status': 'idle',
                                 'allow_auto_upload': True,
                                 'obfuscate_host_names': True}}}


{'generating': {'process': 'Exit Code: pid 35823 exit 0',
                'restart': 'Restart',
                'terminal': 'skipped'},
 'status': 'idle',
 'allow_auto_upload': True,
 'obfuscate_host_names': True}
/var/tmp/report_for_1.tar.gz
{'generating': {'process': 'Exit Code: pid 36482 exit 0',
                'restart': 'Restart',
                'terminal': 'skipped'},
 'status': 'idle',
 'allow_auto_upload': False,
 'obfuscate_host_names': False}
 .2020-07-09 22:10:20 - robottelo - DEBUG - Finished Test: test_cloudinventory.py/test_cloud
```

lhellebr commented:
LGTM, but Travis doesn't like it.

Also, I suggest creating a RFE requesting UI elements to have their ID.

mirekdlugosz commented:
> LGTM, but Travis doesn't like it.

Unrelated to this PR. Sphinx 3.1.1 seems to be more strict that 3.0. See #498 for example of PR where Travis passed.
Maybe we should pin optional requirements, I don't know.

> Also, I suggest creating a RFE requesting UI elements to have their ID.

Could you elaborate? Which elements do you think could use some ids?
Plugin repo is [theforeman/foreman_rh_cloud](https://github.com/theforeman/foreman_rh_cloud/) and I imagine something as simple as adding ids we could do ourselves.

lhellebr commented:
> > Also, I suggest creating a RFE requesting UI elements to have their ID.
> 
> Could you elaborate? Which elements do you think could use some ids?
> Plugin repo is [theforeman/foreman_rh_cloud](https://github.com/theforeman/foreman_rh_cloud/) and I imagine something as simple as adding ids we could do ourselves.

All the elements where you are forced to use xpath instead of id (which is a lot in this PR). From my first encounter with Airgun, the narrative has always been that we want to avoid xpath at all costs, it's one of the main reasons to remake (historical) ui framework and that developers should mark "everything" with id.

omkarkhatavkar commented on airgun/views/cloud_inventory.py:
```diff
@@ -0,0 +1,162 @@
+    @property
+    def is_generating(self):
+        try:
+            self.parent_browser.selenium.find_element_by_xpath(f"{self.STATUS_ELEMENTS}[1]/div")
```
why we are using the `self.parent_browser.selenium.find_element_by_xpath(f"{self.STATUS_ELEMENTS}[1]/div")` is it not possible to get that in any widget ?

mirekdlugosz replied:
Unfortunately, no.

This xpath catches "spinner" ("loading wheel") element, that is only present when work is being done. I want immediate information whether element is present or not, as this function is used by `wait_for` to decide whether we can continue interacting with page, or not.

The way widgetastic is written, all functions I could possibly use (`is_displayed`, `wait_for_element`, `element`), under the hood will call `elements()` recursively. While [`elements()` can take `check_safe=False`](https://github.com/RedHatQE/widgetastic.core/blob/master/src/widgetastic/browser.py#L222) argument to force return value sooner, even on not fully loaded page, it drops this argument in one of recursive calls. In other words, even if I explicitly tell widgetastic to not call `ensure_page_safe()`, it will call it anyway.

And our `ensure_page_safe()` will wait until all spinners on page disappear. But will wait only 30 seconds, without a way for caller to change timeout.

In other words, I know this spinner is there and will remain for some time (in my testing, at least for a minute, but that depends on number of registered hosts), and I have my own wait_for with longer timeout to wait for it to disappear. But under the hood, widgetastic would call default `ensure_page_safe()` with default timeout, which would throw exception before process finishes.

Possible fixes could be:

* change our `ensure_page_safe` to not include spinners, but I have no idea how many existing tests depend on this behavior
* fix widgetastic
* circumvent widgetastic and reach out to Selenium directly

I decided last option will be quickest to implement. Especially considering that I will want that change backported to older Satellite versions, where I can't update widgetastic easily anyway.

omkarkhatavkar replied:
Do you think we need to raise an issue for this in Airgun for suture reference? Okay, I can understand it is a quick and easy fix for now.

mirekdlugosz replied:
It's probably right thing to do, but so low on my priorities list, I don't see myself doing that.
6.7.z, 6.6.z and 6.5.z are not going to benefit from widgetastic core change anyway.

omkarkhatavkar replied:
okay

mirekdlugosz commented:
Changes compared to previous version:

* Added support for new toggle, "Obfuscate host names"
* `InventoryBootstrapSwitch` obtained `class_name` parameter, which defines toggle wrapper class name. It's needed for above change.
* `session.cloudinventory.read()` (without entity name) now correctly reads data from all existing organizations
* slight change to search widget, so `searchbox.clear()` on this page can click clear button and display all organizations

Updated test and test results in initial comment.

mirekdlugosz commented:
@lhellebr 
> All the elements where you are forced to use xpath instead of id (which is a lot in this PR). From my first encounter with Airgun, the narrative has always been that we want to avoid xpath at all costs, it's one of the main reasons to remake (historical) ui framework and that developers should mark "everything" with id.

Obviously I can't comment on what original maintainers said, but isn't it a little more complex than that?
If you look at `airgun/widgets.py`, there's plenty of xpaths. Entire widgetastic is built on xpaths.

I think the intent is more of "no xpaths in tests", "no xpaths in entities" and/or "no xpaths in views". When you instantiate widget, you should prefer name or id, if it's available.

From that point of view, new view is pretty consistent with all the other views we have, keeping xpaths to minimum. xpaths are all hidden inside widgets definitions.

I **could** put these widgets in `airgun/widgets.py`, but since it's the only place that uses them, I decided to keep them close to views. I agree it might raise an eyebrow, as we don't do that often, but we did similar thing in the past - see e.g. `airgun/views/subscription.py`.

Let me know if that addresses your concerns.

lhellebr commented:
My ID-comment is not blocking, feel free to merge. Nevertheless, I think that us being forced to use XPATH instead of just ID can be considered a minor porduct bug and should be discussed with devels. For maintainability sake if for no other reason.
